### ISSUE TITLE
Wrong error about the size of a parameter not being known at compiletime

### ISSUE THREAD

**Issue by DutchGhost** [rust-lang/rust#78834]
I tried this code:

```rust
struct B<A: Sized = [(); {
    let x = [0u8; !0usize];
    1
}]> {
    a: A
}

fn main() {

}
```

I expected to see this happen: This code should compile, as `A` is a generic type with the default type being `[(); 1]`. When I remove the `let x = [0u8; !0usize]` line, everything compiles fine.

Instead, this happened: I got an error message, `error[E0277]`

### Meta

`rustc --version --verbose`:
```
Stable version: 1.47.0
Beta version: 1.48.0-beta.5
1.49.0-nightly (2020-11-05 9d78d1d02761b906038b)
```

<details><summary>Backtrace</summary>
<p>

```
   Compiling playground v0.0.1 (/playground)
error[E0277]: the size for values of type `A` cannot be known at compilation time
  --> src/main.rs:32:19
   |
31 | struct B<A: Sized = [(); {
   |          -
   |          |
   |          this type parameter needs to be `Sized`
   |          required by this bound in `B`
32 |     let x = [0u8; !0usize];
   |                   ^^^^^^^ doesn't have a size known at compile-time

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

</p>
</details>

**Comment by Dylan-DPC** [comment]
Current output:
```
error[[E0277]](https://doc.rust-lang.org/stable/error_codes/E0277.html): the size for values of type `A` cannot be known at compilation time
 --> src/main.rs:2:19
  |
1 |   struct B<A: Sized = [(); {
  |  __________-
2 | |     let x = [0u8; !0usize];
  | |                   ^^^^^^^ doesn't have a size known at compile-time
3 | |     1
4 | | }]> {
  | |__- this type parameter needs to be `std::marker::Sized`
  |
note: required by a bound in `B`
 --> src/main.rs:1:10
  |
1 |   struct B<A: Sized = [(); {
  |  __________^
2 | |     let x = [0u8; !0usize];
3 | |     1
4 | | }]> {
  | |__^ required by this bound in `B`

For more information about this error, try `rustc --explain E0277`.
```

**Comment by fmease** [comment]
If you opt out of `Sized` for the type parameter `A`, then rustc emits the correct error (I assume it's the correct one):

```rs
struct B<
    A: ?Sized = [(); {
                    let x = [0u8; !0usize];
                    1
                }],
> {
    a: A,
}

fn main() {}

```

```
error[E0080]: values of the type `[u8; usize::MAX]` are too big for the current architecture
 --> src/main.rs:3:29
  |
3 |                     let x = [0u8; !0usize];
  |                             ^^^^^^^^^^^^^^
```

**Comment by estebank** [comment]
[Current output](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7692803d8a9ee26bb2a3176c89d21287):

```
error[E0080]: evaluation of constant value failed
 --> src/main.rs:2:13
  |
2 |     let x = [0u8; !0usize];
  |             ^^^^^^^^^^^^^^ values of the type `[u8; usize::MAX]` are too big for the target architecture
```

[and](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=49e5e4f33ebfd4f400d1a135f4ce4826)

```
error[E0080]: evaluation of constant value failed
 --> src/main.rs:3:29
  |
3 |                     let x = [0u8; !0usize];
  |                             ^^^^^^^^^^^^^^ values of the type `[u8; usize::MAX]` are too big for the target architecture
```

**Comment by HypheX** [comment]
@rustbot claim
Thanks, I'll work on this shortly.
